### DISCUSSION TITLE
How to assign default values when getting data elements from an object?

### DISCUSSION THREAD

**Discussion by Enet4** [Enet4/dicom-rs#447] [Q&A]
_Originally posted by @alanwilter in https://github.com/Enet4/dicom-rs/issues/446#issuecomment-1863501964_

> BTW, I'm a python guy and sorry if I'm using the wrong channel (if you know so, please tell me), but how do I do to change `extract_dicom_data()` function so if element is not found (for PatientID, ImageLaterality, etc.) I still want to assign a empty string value in the end. For ImageLaterality actually, I search for Laterality as well and only defaulting to "" if neither were found (this is how I do with `pydicom`).
Now that I put my code to run I'm seeing lots of `No such data element ImageLaterality (with tag (0020,0062))` but I still want to fill my output CSV files with empty results if needed.

```rust
use dicom::dictionary_std::tags;
use dicom::object::OpenFileOptions;
...
fn extract_dicom_data(path: &Path) -> Result<DicomData, Box<dyn std::error::Error>> {
    let obj = OpenFileOptions::new()
        .read_until(tags::PIXEL_DATA)
        .open_file(path)?;

    let patient_id = obj.element_by_name("PatientID")?.to_str()?;
    let image_laterality = obj.element_by_name("ImageLaterality")?.to_str()?;
    let patient_sex = obj.element_by_name("PatientSex")?.to_str()?;
    let patient_dob = obj.element_by_name("PatientBirthDate")?.to_str()?;
    let content_date = obj.element_by_name("ContentDate")?.to_str()?;
    let modality = obj.element_by_name("Modality")?.to_str()?;
    let manufacturer = obj.element_by_name("Manufacturer")?.to_str()?;

    let formatted_patient_dob = format_dicom_date(&patient_dob)?;
    let formatted_content_date = format_dicom_date(&content_date)?;

    let file_path = path
        .canonicalize()?
        .to_str()
        .ok_or("Invalid file path")?
        .to_string();

    Ok(DicomData {
        patient_id: patient_id.to_string(),
        image_laterality: image_laterality.to_string(),
        patient_sex: patient_sex.to_string(),
        patient_dob: formatted_patient_dob,
        content_date: formatted_content_date,
        modality: modality.to_string(),
        manufacturer: manufacturer.to_string(),
        file_path,
    })
}
```

**Comment by Enet4** [comment]
Prefer using `element_opt`, so that you can handle missing elements more effectively.

```rust
use dicom_dictionary_std::tags;

let image_laterality = if let Some(elem) = obj.element_opt(tags::IMAGE_LATERALITY)? {
    elem.to_str()?
} else {
    "".into()
};
```

Note also the preference in the use of methods `element`, `element_opt`, or `get`, rather than `element_by_name`. The latter requires a data element dictionary lookup to resolve the keyword to a tag, which is not necessary because we know the attributes to retrieve at compile time.

**Comment by alanwilter** [comment]
Many thanks for that, this is the perfect place.
I did look into the docs but even that I'm learning how to fetch the right info.
For now, many of my dcms are failing with `premature end of input`.

I'm trying to process in parallel, though I don't see any speedup TBH, and I thought this could be the reason for the `premature end of input` but it is not AFAI can say.

Some details of my code:

```rust
fn main() -> Result<(), Box<dyn std::error::Error>> {
...
    // Process DICOM files in parallel
    let results: Vec<_> = dicom_files
        .par_chunks(num_jobs)
        .map(|chunk| process_dicom_files(chunk))
        .flatten()
        .collect();
...
    Ok(())
}
fn process_dicom_files(paths: &[PathBuf]) -> Vec<DicomData> {
    paths
        .iter()
        .filter_map(|path| match extract_dicom_data(path) {
            Ok(data) => Some(data),
            Err(e) => {
                eprintln!("Error processing {:?}: {}", path, e);
                None
            }
        })
        .collect()
}
```
I've updated `extract_dicom_data()` to use your suggested `obj.element_opt()`.
Note that I do use:
```rust
fn extract_dicom_data(path: &Path) -> Result<DicomData, Box<dyn std::error::Error>> {
    let obj = OpenFileOptions::new()
        .read_until(tags::PIXEL_DATA)
        .open_file(path)?;
```
To avoid reading `PIXEL_DATA` as it makes a huge difference in performance. I thought it could be some dcms not having `PIXEL_DATA`, but that's not the cause either AFAI see. I used `dicom-dump` for some of the dcms that failed and I can clearly see the data looks all fine (BTW, my python code works fine with them).

I use `vscode` and I will look into how to use `debug` for rust projects.
